docs/bigquery_snippets.py

```diff
@@ -0,0 +1,198 @@
+# Copyright 2016 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Testable usage examples for Google Cloud BigQuery API wrapper
+
+Each example function takes a ``client`` argument (which must be an instance
+of :class:`gcloud.bigquery.client.Client`) and uses it to perform a task with
+the API.
+
+To facility running the examples as system tests, each example is also passed
+a ``to_delete`` list;  the function adds to the list any objects created which
+need to be deleted during teardown.
+"""
+
+import time
+
+from gcloud.bigquery.client import Client
+
+
+def snippet(func):
+    """Mark ``func`` as a snippet example function."""
+    func._snippet = True
+    return func
+
+
+def _millis():
+    return time.time() * 1000
+
+
+@snippet
+def client_list_datasets(client, to_delete):  # pylint: disable=unused-argument
+    """List datasets for a project."""
+
+    def do_something_with(sub):  # pylint: disable=unused-argument
+        pass
+
+    # [START client_list_datasets]
+    datasets, token = client.list_datasets()   # API request
+    while True:
+        for dataset in datasets:
+            do_something_with(dataset)
+        if token is None:
+            break
+        datasets, token = client.list_datasets(page_token=token)  # API request
+    # [END client_list_datasets]
+
+
+@snippet
+def dataset_create(client, to_delete):
+    """Create a dataset."""
+    DATASET_NAME = 'dataset_create_%d' % (_millis(),)
+
+    # [START dataset_create]
+    dataset = client.dataset(DATASET_NAME)
+    dataset.create()              # API request
+    # [END dataset_create]
+
+    to_delete.append(dataset)
+
+
+@snippet
+def dataset_exists(client, to_delete):
+    """Test existence of a dataset."""
+    DATASET_NAME = 'dataset_exists_%d' % (_millis(),)
+    dataset = client.dataset(DATASET_NAME)
+    to_delete.append(dataset)
+
+    # [START dataset_exists]
+    assert not dataset.exists()   # API request
+    dataset.create()              # API request
+    assert dataset.exists()       # API request
```

Comment:
I left them outside in cases where they weren't important to show the semantics of the method being documented.  Here, they help the user understand what happens.